esrally/storage/_client.py

```diff
@@ -0,0 +1,220 @@
+        # LOG.debug("weights: %s", weights)
+        mirror_urls.sort(key=lambda u: weights[u])
+        # LOG.debug("mirror_urls: %s", mirror_urls)
+
+        for u in mirror_urls:
+            try:
+                head = self.head(u, ttl=ttl)
+            except FileNotFoundError:
+                LOG.debug("file not found: %r", url)
+                continue
+            except Exception as ex:
+                LOG.debug("error getting file head: %r, %s", url, ex)
+                continue
+            if content_length is not None and content_length != head.content_length:
+                LOG.debug("unexpected content length: %r, got %d, want %d", url, content_length, head.content_length)
+                continue
+            if accept_ranges and not head.accept_ranges:
+                LOG.debug("it doesn't accept ranges: %r", url)
+                continue
+            yield head
+
+    def get(self, url: str, stream: Writable, ranges: RangeSet = NO_RANGE) -> Head:
+        """It downloads a remote bucket object to a local file path.
+
+        :param url: it represents the URL of the remote file.
+        :param stream: it represents the destination file stream where to write data to.
+        :param ranges: it represents the portion of the file to transfer.
+        :raises ServiceUnavailableError: in case on temporary service failure.
+        """
+        for head in self.resolve(url, accept_ranges=bool(ranges)):
+            connections = self._server_connections(head.url)
+            try:
+                connections.add(1)
+            except WaitGroupLimitError:
+                continue
+            adapter = self._adapters.get(head.url)
+            try:
+                return adapter.get(head.url, stream, ranges)
+            except ServiceUnavailableError:
+                with self._lock:
+                    connections.max_count = max(1, connections.count - 1)
+            finally:
+                connections.done()
+
+        raise ServiceUnavailableError(f"any service available for getting {url}")
+
+    def _server_connections(self, url: str) -> WaitGroup:
+        with self._lock:
+            return self._connections[_server_key(url)]
+
+    def _server_stats_from_url(self, url: str, ttl: float | None = None) -> deque[ServerStats]:
+        with self._lock:
+            stats = self._stats[_server_key(url)]
+            if ttl is not None:
+                # it removes expired latencies
+                deadline = time.monotonic() - ttl
+                while stats and stats[0].end < deadline:
+                    stats.popleft()
+        return stats
+
+    def _average_latency(self, url: str, ttl: float | None = 60.0) -> float:
+        stats = self._server_stats_from_url(url, ttl)
+        if not stats:
+            return 0.0
+        return sum(s.latency for s in stats) / len(stats)
+
+    def monitor(self):
```

Comment:
Should there be a `with self._lock:` here? I'm worried about a race condition, but I suppose it'd just be old data if we do end up calling another thread during this. And of course until the GIL is gone, we won't really have any issues, anyway.